Generics in trait definition

Hi, I was experimenting with dynosaur and ran into an issue while trying to create my own `Service` trait.

```rust
#[dynosaur::dynosaur(DynService)]
pub trait Service<Request> {
    type Response;
    type Error;

    fn call(&self, req: Request) -> impl Future<Output = Result<Self::Response, Self::Error>>;
}
```
This definition doesn't compile and returns this error:
```
error[E0309]: the parameter type `Request` may not live long enough
 --> src/lib.rs:3:1
  |
3 | #[dynosaur::dynosaur(DynService)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | the parameter type `Request` must be valid for the lifetime `'dynosaur` as defined here...
  | ...so that the type `Request` will meet its required lifetime bounds
  |
  = note: this error originates in the attribute macro `dynosaur::dynosaur` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider adding an explicit lifetime bound
  |
3 | #[dynosaur::dynosaur(DynService)], Request: 'dynosaur
  |                                  ++++++++++++++++++++
```
After expanding the code with `cargo expand`, I added the suggested `Request: 'dynosaur` bound in the generated `ErasedService` which seams to solve the issue.

--- a/dynosaur/tests/pass/generics-in-trait-def.rs
+++ b/dynosaur/tests/pass/generics-in-trait-def.rs
@@ -1,0 +1,11 @@
+use std::future::Future;
+
+#[dynosaur::dynosaur(DynService)]
+pub trait Service<Request> {
+    type Response;
+    type Error;
+
+    fn call(&self, req: Request) -> impl Future<Output = Result<Self::Response, Self::Error>>;
+}
+
+fn main() {}

--- a/dynosaur/tests/pass/generics-in-trait-def.stdout
+++ b/dynosaur/tests/pass/generics-in-trait-def.stdout
@@ -1,0 +1,148 @@
+#![feature(prelude_import)]
+#[prelude_import]
+use std::prelude::rust_2021::*;
+#[macro_use]
+extern crate std;
+use std::future::Future;
+
+pub trait Service<Request> {
+    type Response;
+    type Error;
+
+    fn call(&self, req: Request)
+    -> impl Future<Output = Result<Self::Response, Self::Error>>;
+}
+mod __dynosaur_macro_dynservice {
+    use super::*;
+    pub trait ErasedService<Request> {
+        type Response;
+        type Error;
+        fn call<'life0, 'dynosaur>(&'life0 self, req: Request)
+        ->
+            ::core::pin::Pin<Box<dyn Future<Output =
+            Result<Self::Response, Self::Error>> + 'dynosaur>>
+        where
+        Request: 'dynosaur,
+        'life0: 'dynosaur,
+        Self: 'dynosaur;
+    }
+    impl<Request, DYNOSAUR: Service<Request>> ErasedService<Request> for
+        DYNOSAUR {
+        type Response = <Self as Service<Request>>::Response;
+        type Error = <Self as Service<Request>>::Error;
+        fn call<'life0, 'dynosaur>(&'life0 self, req: Request)
+            ->
+                ::core::pin::Pin<Box<dyn Future<Output =
+                Result<Self::Response, Self::Error>> + 'dynosaur>> where
+            Request: 'dynosaur, 'life0: 'dynosaur, Self: 'dynosaur {
+            Box::pin(<Self as Service<Request>>::call(self, req))
+        }
+    }
+    #[repr(transparent)]
+    pub struct DynService<'dynosaur_struct, Request, Response, Error> {
+        ptr: dyn ErasedService<Request, Response = Response, Error = Error> +
+        'dynosaur_struct,
+    }
+    impl<'dynosaur_struct, Request, Response, Error> Service<Request> for
+        DynService<'dynosaur_struct, Request, Response, Error> {
+        type Response = Response;
+        type Error = Error;
+        fn call(&self, req: Request)
+            -> impl Future<Output = Result<Self::Response, Self::Error>> {
+            let ret:
+                    ::core::pin::Pin<Box<dyn Future<Output =
+                    Result<Self::Response, Self::Error>> + '_>> =
+                self.ptr.call(req);
+            let ret:
+                    ::core::pin::Pin<Box<dyn Future<Output =
+                    Result<Self::Response, Self::Error>> + 'static>> =
+                unsafe { ::core::mem::transmute(ret) };
+            ret
+        }
+    }
+    impl<'dynosaur_struct, Request, Response, Error>
+        DynService<'dynosaur_struct, Request, Response, Error> {
+        pub fn new_box(value:
+                impl Service<Request, Response = Response, Error = Error> +
+                'dynosaur_struct)
+            -> Box<DynService<'dynosaur_struct, Request, Response, Error>> {
+            let value = Box::new(value);
+            let value:
+                    Box<dyn ErasedService<Request, Response = Response, Error =
+                    Error> + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn new_arc(value:
+                impl Service<Request, Response = Response, Error = Error> +
+                'dynosaur_struct)
+            ->
+                std::sync::Arc<DynService<'dynosaur_struct, Request, Response,
+                Error>> {
+            let value = std::sync::Arc::new(value);
+            let value:
+                    std::sync::Arc<dyn ErasedService<Request, Response =
+                    Response, Error = Error> + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn new_rc(value:
+                impl Service<Request, Response = Response, Error = Error> +
+                'dynosaur_struct)
+            ->
+                std::rc::Rc<DynService<'dynosaur_struct, Request, Response,
+                Error>> {
+            let value = std::rc::Rc::new(value);
+            let value:
+                    std::rc::Rc<dyn ErasedService<Request, Response = Response,
+                    Error = Error> + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub const fn from_ref(value:
+                &(impl Service<Request, Response = Response, Error = Error> +
+                'dynosaur_struct))
+            -> &DynService<'dynosaur_struct, Request, Response, Error> {
+            let value:
+                    &(dyn ErasedService<Request, Response = Response, Error =
+                    Error> + 'dynosaur_struct) = &*value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub const fn from_mut(value:
+                &mut (impl Service<Request, Response = Response, Error =
+                Error> + 'dynosaur_struct))
+            -> &mut DynService<'dynosaur_struct, Request, Response, Error> {
+            let value:
+                    &mut (dyn ErasedService<Request, Response = Response, Error
+                    = Error> + 'dynosaur_struct) = &mut *value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+    }
+    impl<Request, DYNOSAUR: Service<Request>> Service<Request> for &DYNOSAUR
+        where DYNOSAUR: ?Sized {
+        type Response = DYNOSAUR::Response;
+        type Error = DYNOSAUR::Error;
+        fn call(&self, req: Request)
+            -> impl Future<Output = Result<Self::Response, Self::Error>> {
+            DYNOSAUR::call(self, req)
+        }
+    }
+    impl<Request, DYNOSAUR: Service<Request>> Service<Request> for
+        &mut DYNOSAUR where DYNOSAUR: ?Sized {
+        type Response = DYNOSAUR::Response;
+        type Error = DYNOSAUR::Error;
+        fn call(&self, req: Request)
+            -> impl Future<Output = Result<Self::Response, Self::Error>> {
+            DYNOSAUR::call(self, req)
+        }
+    }
+    impl<Request, DYNOSAUR: Service<Request>> Service<Request> for
+        Box<DYNOSAUR> where DYNOSAUR: ?Sized {
+        type Response = DYNOSAUR::Response;
+        type Error = DYNOSAUR::Error;
+        fn call(&self, req: Request)
+            -> impl Future<Output = Result<Self::Response, Self::Error>> {
+            DYNOSAUR::call(self, req)
+        }
+    }
+}
+use __dynosaur_macro_dynservice::DynService;
+
+fn main() {}

--- a/dynosaur/tests/pass/multiple-lifetimes-and-where-clauses.stdout
+++ b/dynosaur/tests/pass/multiple-lifetimes-and-where-clauses.stdout
@@ -20,11 +20,11 @@ mod __dynosaur_macro_dynsometrait {
             &'a u32> + 'dynosaur>>
         where
         'b: 'a,
+        'a: 'dynosaur,
+        'b: 'dynosaur,
         'd: 'dynosaur,
         'e: 'dynosaur,
         'f: 'dynosaur,
-        'a: 'dynosaur,
-        'b: 'dynosaur,
         'life0: 'dynosaur,
         Self: 'dynosaur;
     }
@@ -34,8 +34,8 @@ mod __dynosaur_macro_dynsometrait {
             'dynosaur>(&'life0 self, a: &'d u32, b: &'e u32, c: &'f u32)
             ->
                 ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
-                &'a u32> + 'dynosaur>> where 'b: 'a, 'd: 'dynosaur,
-            'e: 'dynosaur, 'f: 'dynosaur, 'a: 'dynosaur, 'b: 'dynosaur,
+                &'a u32> + 'dynosaur>> where 'b: 'a, 'a: 'dynosaur,
+            'b: 'dynosaur, 'd: 'dynosaur, 'e: 'dynosaur, 'f: 'dynosaur,
             'life0: 'dynosaur, Self: 'dynosaur {
             Box::pin(<Self as
                         SomeTrait<'a, 'b>>::lotsa_lifetimes(self, a, b, c))

--- a/dynosaur/tests/pass/multiple-lifetimes.rs
+++ b/dynosaur/tests/pass/multiple-lifetimes.rs
@@ -1,14 +1,14 @@
 #[dynosaur::dynosaur(DynSomeTrait)]
 trait SomeTrait {
     async fn multiple_elided_lifetimes(&self, a: &u8, b: &u8);
-    async fn multiple_named_lifetimes<'a, 'b>(&self, a: &'a u8, b: &'b u8, c: &u8);
+    async fn multiple_named_lifetimes<'a, 'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8);
     async fn same_lifetimes_multiple_params<'a>(&self, a1: &'a u8, a2: &'a u8, c: &u8) -> &'a u8;
     async fn multiple_static_and_anonymous(&self, a: &'static u8, b: &'_ u8, c: &'_ u8);
 }
 
 impl SomeTrait for () {
     async fn multiple_elided_lifetimes(&self, a: &u8, b: &u8) {}
-    async fn multiple_named_lifetimes<'a, 'b>(&self, a: &'a u8, b: &'b u8, c: &u8) {}
+    async fn multiple_named_lifetimes<'a, 'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8) {}
     async fn same_lifetimes_multiple_params<'a>(&self, a1: &'a u8, a2: &'a u8, c: &u8) -> &'a u8 {
         a2
     }

--- a/dynosaur/tests/pass/multiple-lifetimes.stdout
+++ b/dynosaur/tests/pass/multiple-lifetimes.stdout
@@ -26,7 +26,7 @@ mod __dynosaur_macro_dynsometrait {
         'life1: 'dynosaur,
         'life2: 'dynosaur,
         Self: 'dynosaur;
-        fn multiple_named_lifetimes<'a, 'b: 'a, 'life0, 'life1,
+        fn multiple_named_lifetimes<'a, 'b, 'life0, 'life1,
         'dynosaur>(&'life0 self, a: &'a u8, b: &'b u8, c: &'life1 u8)
         ->
             ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
@@ -68,7 +68,7 @@ mod __dynosaur_macro_dynsometrait {
             Box::pin(<Self as
                         SomeTrait>::multiple_elided_lifetimes(self, a, b))
         }
-        fn multiple_named_lifetimes<'a, 'b: 'a, 'life0, 'life1,
+        fn multiple_named_lifetimes<'a, 'b, 'life0, 'life1,
             'dynosaur>(&'life0 self, a: &'a u8, b: &'b u8, c: &'life1 u8)
             ->
                 ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +

--- a/dynosaur/tests/pass/multiple-lifetimes.stdout
+++ b/dynosaur/tests/pass/multiple-lifetimes.stdout
@@ -6,7 +6,7 @@ extern crate std;
 trait SomeTrait {
     async fn multiple_elided_lifetimes(&self, a: &u8, b: &u8);
     async fn multiple_named_lifetimes<'a,
-    'b>(&self, a: &'a u8, b: &'b u8, c: &u8);
+    'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8);
     async fn same_lifetimes_multiple_params<'a>(&self, a1: &'a u8, a2: &'a u8,
     c: &u8)
     -> &'a u8;
@@ -33,7 +33,7 @@ mod __dynosaur_macro_dynsometrait {
             'dynosaur>>
         where
         'a: 'dynosaur,
-        'b: 'dynosaur,
+        'b: 'dynosaur + 'a,
         'life0: 'dynosaur,
         'life1: 'dynosaur,
         Self: 'dynosaur;
@@ -72,7 +72,7 @@ mod __dynosaur_macro_dynsometrait {
             'dynosaur>(&'life0 self, a: &'a u8, b: &'b u8, c: &'life1 u8)
             ->
                 ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
-                'dynosaur>> where 'a: 'dynosaur, 'b: 'dynosaur,
+                'dynosaur>> where 'a: 'dynosaur, 'b: 'dynosaur + 'a,
             'life0: 'dynosaur, 'life1: 'dynosaur, Self: 'dynosaur {
             Box::pin(<Self as
                         SomeTrait>::multiple_named_lifetimes(self, a, b, c))
@@ -113,7 +113,7 @@ mod __dynosaur_macro_dynsometrait {
             ret
         }
         fn multiple_named_lifetimes<'a,
-            'b>(&self, a: &'a u8, b: &'b u8, c: &u8)
+            'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8)
             -> impl ::core::future::Future<Output = ()> {
             let ret:
                     ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
@@ -187,7 +187,7 @@ mod __dynosaur_macro_dynsometrait {
             DYNOSAUR::multiple_elided_lifetimes(self, a, b)
         }
         fn multiple_named_lifetimes<'a,
-            'b>(&self, a: &'a u8, b: &'b u8, c: &u8)
+            'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8)
             -> impl ::core::future::Future<Output = ()> {
             DYNOSAUR::multiple_named_lifetimes(self, a, b, c)
         }
@@ -207,7 +207,7 @@ mod __dynosaur_macro_dynsometrait {
             DYNOSAUR::multiple_elided_lifetimes(self, a, b)
         }
         fn multiple_named_lifetimes<'a,
-            'b>(&self, a: &'a u8, b: &'b u8, c: &u8)
+            'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8)
             -> impl ::core::future::Future<Output = ()> {
             DYNOSAUR::multiple_named_lifetimes(self, a, b, c)
         }
@@ -227,7 +227,7 @@ mod __dynosaur_macro_dynsometrait {
             DYNOSAUR::multiple_elided_lifetimes(self, a, b)
         }
         fn multiple_named_lifetimes<'a,
-            'b>(&self, a: &'a u8, b: &'b u8, c: &u8)
+            'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8)
             -> impl ::core::future::Future<Output = ()> {
             DYNOSAUR::multiple_named_lifetimes(self, a, b, c)
         }
@@ -246,7 +246,7 @@ use __dynosaur_macro_dynsometrait::DynSomeTrait;
 impl SomeTrait for () {
     async fn multiple_elided_lifetimes(&self, a: &u8, b: &u8) {}
     async fn multiple_named_lifetimes<'a,
-        'b>(&self, a: &'a u8, b: &'b u8, c: &u8) {}
+        'b: 'a>(&self, a: &'a u8, b: &'b u8, c: &u8) {}
     async fn same_lifetimes_multiple_params<'a>(&self, a1: &'a u8, a2: &'a u8,
         c: &u8) -> &'a u8 {
         a2

--- a/dynosaur/tests/pass/multiple-lifetimes.stdout
+++ b/dynosaur/tests/pass/multiple-lifetimes.stdout
@@ -26,14 +26,14 @@ mod __dynosaur_macro_dynsometrait {
         'life1: 'dynosaur,
         'life2: 'dynosaur,
         Self: 'dynosaur;
-        fn multiple_named_lifetimes<'a, 'b, 'life0, 'life1,
+        fn multiple_named_lifetimes<'a, 'b: 'a, 'life0, 'life1,
         'dynosaur>(&'life0 self, a: &'a u8, b: &'b u8, c: &'life1 u8)
         ->
             ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
             'dynosaur>>
         where
         'a: 'dynosaur,
-        'b: 'dynosaur + 'a,
+        'b: 'a + 'dynosaur,
         'life0: 'dynosaur,
         'life1: 'dynosaur,
         Self: 'dynosaur;
@@ -68,11 +68,11 @@ mod __dynosaur_macro_dynsometrait {
             Box::pin(<Self as
                         SomeTrait>::multiple_elided_lifetimes(self, a, b))
         }
-        fn multiple_named_lifetimes<'a, 'b, 'life0, 'life1,
+        fn multiple_named_lifetimes<'a, 'b: 'a, 'life0, 'life1,
             'dynosaur>(&'life0 self, a: &'a u8, b: &'b u8, c: &'life1 u8)
             ->
                 ::core::pin::Pin<Box<dyn ::core::future::Future<Output = ()> +
-                'dynosaur>> where 'a: 'dynosaur, 'b: 'dynosaur + 'a,
+                'dynosaur>> where 'a: 'dynosaur, 'b: 'a + 'dynosaur,
             'life0: 'dynosaur, 'life1: 'dynosaur, Self: 'dynosaur {
             Box::pin(<Self as
                         SomeTrait>::multiple_named_lifetimes(self, a, b, c))

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -105,18 +105,32 @@ fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {
         }
     }
 
-    for lifetime in item_trait_generics
-        .params
-        .iter()
-        .filter_map(move |param| match param {
-            GenericParam::Lifetime(param) => Some(&param.lifetime),
-            _ => None,
-        })
-    {
-        let span = lifetime.span();
-        where_clause_or_default(&mut sig.generics.where_clause)
-            .predicates
-            .push(parse_quote_spanned!(span=> #lifetime: 'dynosaur));
+    for param in &item_trait_generics.params {
+        match param {
+            GenericParam::Type(param) => {
+                let param_name = &param.ident;
+                let span = match param.colon_token {
+                    Some(colon_token) => colon_token.span,
+                    None => param_name.span(),
+                };
+                let bounds = &param.bounds;
+                where_clause_or_default(&mut sig.generics.where_clause)
+                    .predicates
+                    .push(parse_quote_spanned!(span=> #param_name: 'dynosaur + #bounds));
+            }
+            GenericParam::Lifetime(param) => {
+                let param_name = &param.lifetime;
+                let span = match param.colon_token {
+                    Some(colon_token) => colon_token.span,
+                    None => param_name.span(),
+                };
+                let bounds = &param.bounds;
+                where_clause_or_default(&mut sig.generics.where_clause)
+                    .predicates
+                    .push(parse_quote_spanned!(span=> #param: 'dynosaur + #bounds));
+            }
+            GenericParam::Const(_) => {}
+        }
     }
 
     if sig.generics.lt_token.is_none() {

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -4,12 +4,12 @@ use crate::sig::{is_async, is_future, is_rpit};
 use crate::where_clauses::{has_where_self_sized, where_clause_or_default};
 use proc_macro2::{Span, TokenStream};
 use quote::quote;
-use std::mem;
 use syn::punctuated::Punctuated;
 use syn::visit_mut::VisitMut;
 use syn::{
     parse_quote, parse_quote_spanned, Error, FnArg, GenericParam, Generics, Ident, ItemTrait,
     Lifetime, Pat, PatIdent, ReturnType, Signature, Token, Type, TypeImplTrait, TypeParamBound,
+    WherePredicate,
 };
 
 /// Expands the signature of each function on the trait, converting async fn into fn with return
@@ -77,35 +77,11 @@ fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {
         }
     }
 
-    for param in &mut sig.generics.params {
-        match param {
-            GenericParam::Type(param) => {
-                let param_name = &param.ident;
-                let span = match param.colon_token.take() {
-                    Some(colon_token) => colon_token.span,
-                    None => param_name.span(),
-                };
-                let bounds = mem::replace(&mut param.bounds, Punctuated::new());
-                where_clause_or_default(&mut sig.generics.where_clause)
-                    .predicates
-                    .push(parse_quote_spanned!(span=> #param_name: 'dynosaur + #bounds));
-            }
-            GenericParam::Lifetime(param) => {
-                let param_name = &param.lifetime;
-                let span = match param.colon_token.take() {
-                    Some(colon_token) => colon_token.span,
-                    None => param_name.span(),
-                };
-                let bounds = mem::replace(&mut param.bounds, Punctuated::new());
-                where_clause_or_default(&mut sig.generics.where_clause)
-                    .predicates
-                    .push(parse_quote_spanned!(span=> #param: 'dynosaur + #bounds));
-            }
-            GenericParam::Const(_) => {}
-        }
-    }
-
-    for param in &item_trait_generics.params {
+    for param in item_trait_generics
+        .params
+        .iter()
+        .chain(sig.generics.params.iter())
+    {
         match param {
             GenericParam::Type(param) => {
                 let param_name = &param.ident;
@@ -113,10 +89,14 @@ fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {
                     Some(colon_token) => colon_token.span,
                     None => param_name.span(),
                 };
-                let bounds = &param.bounds;
+                let mut bounds = param.bounds.clone();
+                bounds.push(TypeParamBound::Lifetime(Lifetime::new(
+                    "'dynosaur",
+                    Span::call_site(),
+                )));
                 where_clause_or_default(&mut sig.generics.where_clause)
                     .predicates
-                    .push(parse_quote_spanned!(span=> #param_name: 'dynosaur + #bounds));
+                    .push(parse_quote_spanned!(span=> #param_name: #bounds));
             }
             GenericParam::Lifetime(param) => {
                 let param_name = &param.lifetime;
@@ -124,10 +104,11 @@ fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {
                     Some(colon_token) => colon_token.span,
                     None => param_name.span(),
                 };
-                let bounds = &param.bounds;
+                let mut bounds = param.bounds.clone();
+                bounds.push(Lifetime::new("'dynosaur", Span::call_site()));
                 where_clause_or_default(&mut sig.generics.where_clause)
                     .predicates
-                    .push(parse_quote_spanned!(span=> #param: 'dynosaur + #bounds));
+                    .push(parse_quote_spanned!(span=> #param_name: #bounds));
             }
             GenericParam::Const(_) => {}
         }

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -114,6 +114,18 @@ fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {
         }
     }
 
+    for param in &mut sig.generics.params {
+        match param {
+            GenericParam::Type(param) => {
+                param.bounds = Punctuated::new();
+            }
+            GenericParam::Lifetime(param) => {
+                param.bounds = Punctuated::new();
+            }
+            GenericParam::Const(_) => {}
+        }
+    }
+
     if sig.generics.lt_token.is_none() {
         sig.generics.lt_token = Some(Token![<](sig.ident.span()));
     }

